## 計算環境
MacBook Pro (13-inch, 2016, Two Thunderbolt 3 ports)

プロセッサ: 2 GHz Intel Core i5

メモリ: 8 GB 1867 MHz LPDDR3

ソフト: Jupyter notebook (5.4.0)

言語: Python 3.6.4

# 大問1

In [1]:
import numpy as np

In [2]:
# 初期化

A_1 = np.array([[1, 0, 0], [0, -5, 1], [0, 3, 0]])
A_2 = np.array([[1, 0, 0], [0, 0, -1], [0, 3, 0]])

v = np.array([[1], [1], [1]])

In [3]:
def power_method(A, x):
    tol = 1e-6
    error = tol + 1
    x_old = x

    while error > tol:
        y = np.dot(A, x_old)
        x = y / np.linalg.norm(y)
        error = min([np.linalg.norm(x - x_old), np.linalg.norm(x + x_old)]) / np.linalg.norm(x)
        print(x - x_old)
        x_old = x
        
    return x

In [4]:
def power_method(A, x):
    for i in range(5000):
        y = np.dot(A, x)
        x = y / np.linalg.norm(y)
        if i > 4990:
            print("=============")
            print(x)

## (A1)

In [5]:
power_method(A_1, v)

[[ 0.        ]
 [ 0.87939705]
 [-0.47608909]]
[[ 0.        ]
 [-0.87939705]
 [ 0.47608909]]
[[ 0.        ]
 [ 0.87939705]
 [-0.47608909]]
[[ 0.        ]
 [-0.87939705]
 [ 0.47608909]]
[[ 0.        ]
 [ 0.87939705]
 [-0.47608909]]
[[ 0.        ]
 [-0.87939705]
 [ 0.47608909]]
[[ 0.        ]
 [ 0.87939705]
 [-0.47608909]]
[[ 0.        ]
 [-0.87939705]
 [ 0.47608909]]
[[ 0.        ]
 [ 0.87939705]
 [-0.47608909]]


In [6]:
np.linalg.norm(np.array([[ 0.        ],  [ 0.87939705], [-0.47608909]]))

0.9999999965828653

In [7]:
np.dot(A_1, np.array([[ 0.        ],  [ 0.87939705], [-0.47608909]])) / np.array([[ 0.        ],  [ 0.87939705], [-0.47608909]])

/Users/uedatomohiro/.pyenv/versions/anaconda3-5.1.0/lib/python3.6/site-packages/ipykernel_launcher.py:1: RuntimeWarning: invalid value encountered in true_divide
  """Entry point for launching an IPython kernel.


array([[        nan],
       [-5.54138127],
       [-5.54138124]])

In [8]:
np.linalg.eig(A_1)

(array([ 0.54138127, -5.54138127,  1.        ]),
 array([[ 0.        ,  0.        ,  1.        ],
        [ 0.17759186,  0.87939705,  0.        ],
        [ 0.98410423, -0.47608909,  0.        ]]))

kが十分大きいところでは上のように、互いに反対向きの大きさ1のベクトルを交互に繰り返している。

このベクトルは

np.linalg.eig(A_1)

や

np.dot(A_1, np.array([[ 0.        ],  [ 0.87939705], [-0.47608909]])) / np.array([[ 0.        ],  [ 0.87939705], [-0.47608909]]) = [[nan], [-5.54128127], [-5.54128127]]

となることからわかるように、A_1の固有ベクトルであり、その固有値は-5.54138127で、固有値の中で絶対値最大である。

## A2

In [9]:
power_method(A_2, v)

[[0.        ]
 [0.70710678]
 [0.70710678]]
[[ 0.        ]
 [-0.31622777]
 [ 0.9486833 ]]
[[ 0.        ]
 [-0.70710678]
 [-0.70710678]]
[[ 0.        ]
 [ 0.31622777]
 [-0.9486833 ]]
[[0.        ]
 [0.70710678]
 [0.70710678]]
[[ 0.        ]
 [-0.31622777]
 [ 0.9486833 ]]
[[ 0.        ]
 [-0.70710678]
 [-0.70710678]]
[[ 0.        ]
 [ 0.31622777]
 [-0.9486833 ]]
[[0.        ]
 [0.70710678]
 [0.70710678]]


In [10]:
np.linalg.eig(A_2)

(array([0.+1.73205081j, 0.-1.73205081j, 1.+0.        j]),
 array([[0.       +0. j, 0.       -0. j, 1.       +0. j],
        [0.       +0.5j, 0.       -0.5j, 0.       +0. j],
        [0.8660254+0. j, 0.8660254-0. j, 0.       +0. j]]))

紙に書きます